examples/structured_data/time_series_anomaly_using_autoencoder.py

```diff
@@ -0,0 +1,323 @@
+"""
+
+master_url_root = "https://raw.githubusercontent.com/numenta/NAB/master/data/"
+
+df_small_noise_url_suffix = "artificialNoAnomaly/art_daily_small_noise.csv"
+df_small_noise_url = master_url_root + df_small_noise_url_suffix
+df_small_noise = pd.read_csv(df_small_noise_url)
+
+df_daily_jumpsup_url_suffix = "artificialWithAnomaly/art_daily_jumpsup.csv"
+df_daily_jumpsup_url = master_url_root + df_daily_jumpsup_url_suffix
+df_daily_jumpsup = pd.read_csv(df_daily_jumpsup_url)
+
+"""
+## Quick look at the data
+"""
+
+print(df_small_noise.head())
+
+print(df_daily_jumpsup.head())
+
+"""
+## Visualize the data
+"""
+
+
+def plot_dates_values(data):
+    dates = data["timestamp"].to_list()
+    values = data["value"].to_list()
+    dates = [datetime.strptime(x, "%Y-%m-%d %H:%M:%S") for x in dates]
+    plt.subplots_adjust(bottom=0.2)
+    plt.xticks(rotation=25)
+    ax = plt.gca()
+    xfmt = md.DateFormatter("%Y-%m-%d %H:%M:%S")
+    ax.xaxis.set_major_formatter(xfmt)
+    plt.plot(dates, values)
+    plt.show()
+
+
+"""
+### Timeseries data without anomalies
+
+We will use the following data for training.
+"""
+
+plot_dates_values(df_small_noise)
+
+"""
+### Timeseries data with anomalies
+
+We will use the following data for testing and see if the sudden jump up in the data is
+detected as an anomaly.
+"""
+
+plot_dates_values(df_daily_jumpsup)
+
+"""
+## Prepare training data
+
+Get data values from the training timeseries data file and normalize the `value` data. We
+have a `value` for every 5 mins for 14 days.
+*   24 * 60 / 5 = **288 timesteps per day**
+*   288 * 14 = **4032 data points** in total
+"""
+
+
+def get_value_from_df(df):
+    return df.value.to_list()
+
+
+def normalize(values):
+    mean = np.mean(values)
+    values -= mean
+    std = np.std(values)
+    values /= std
+    return values, mean, std
+
+
+# Get the `value` column from the training dataframe.
+training_value = get_value_from_df(df_small_noise)
+
+# Normalize `value` and save the mean and std we get, for normalizing test data.
+training_value, training_mean, training_std = normalize(training_value)
+len(training_value)
+
+"""
+### Create sequences
+Create sequences combining `TIME_STEPS` contiguous data values from the training data.
+"""
+
+TIME_STEPS = 288
+
+
+def create_sequences(values, time_steps=TIME_STEPS):
+    output = []
+    for i in range(len(values) - time_steps):
+        output.append(values[i : (i + time_steps)])
+    # Convert 2D sequences into 3D as we will be feeding this into
+    # a convolutional layer.
+    return np.expand_dims(output, axis=2)
+
+
+x_train = create_sequences(training_value)
+print("Training input shape: ", x_train.shape)
+
+"""
+## Build a model
+
+We will build a convolutional reconstruction autoencoder model. The model will take input
+of shape `(batch_size, sequence_length, num_features)` and return output of the same
+shape. In this case, `sequence_length` is 288 and `num_features` is 1.
+"""
+
+model = Sequential(
```

Comment:
I would also recommend adding `relu` activations in the intermediate layers (not the last layer). The lack of non-linearity here is surprising

Reply:
Yes, it was surprising to me as well. Anything implicit Keras is doing that is non-linear? I added relu activations now.

Reply:
Yes, the strides and the mixing across convolution windows as you stack multiple layers make the transformations non-linear. However relu should definitely help with learning.